### DISCUSSION TITLE
ip64-router

### DISCUSSION THREAD

**Discussion by tlzfighting** [contiki-ng/contiki-ng#2334] [Q&A]
###     **I want to implement ip64-router on the cc2538dk platform, so I commented some of the code in Makefile:**
![图片](https://user-images.githubusercontent.com/95647449/206916060-ae892c63-4b7b-48aa-ada5-d3c4e607c228.png)
 
### **And I added the "ip64-conf.h" file under the directory "arch/platform/cc2538dk"**
![图片](https://user-images.githubusercontent.com/95647449/206916154-1ffa109c-04c8-466d-9b71-c77f802b37ce.png)

### **But when I type the compile command "make TARGET=cc2538dk."", the following error occurred.**

../../arch/platform/cc2538dk/./ip64-conf.h:41:10: fatal error: enc28j60-ip64-driver.h: No such file or directory
   41 | #include "enc28j60-ip64-driver.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make: *** [../../Makefile.include:498: build/cc2538dk/obj/ip64-router.o] Error 1

### **How can I solve this problem for you, look forward to your valuable suggestions!**

**Comment by pjonsson** [comment]
The example depends on the enc28j60 ethernet driver, see `arch/platform/zoul/orion/Makefile.orion` for the `MODULES += ...` line that adds the driver and the following lines that sets up the build configuration of the driver.

**Reply by tlzfighting** [reply]
> The example depends on the enc28j60 ethernet driver, see `arch/platform/zoul/orion/Makefile.orion` for the `MODULES += ...` line that adds the driver and the following lines that sets up the build configuration of the driver.


Comment options
tlzfighting
Dec 13, 2022
Author
Hi, sir !

### I added the enc28j60 in the cc2538dk.Makefile.
MODULES += $(CONTIKI_NG_DRIVERS_ETHERNET_DIR)/enc28j60
But when I type the compile command "make TARGET=cc2538dk."", the following error occurred.

ip64-router.c:(.text.process_thread_router_node_process+0x1a): undefined reference to `ip64_init'
collect2: error: ld returned 1 exit status
make: *** [../../arch/cpu/arm/cortex-m/Makefile.customrules-cortex-m:11: build/cc2538dk/ip64-router.elf] Error 1

### How can I solve this problem for you, look forward to your valuable suggestions!

**Reply by pjonsson** [reply]
Does your cc2538dk board have a enc28j60 device?

**Reply by tlzfighting** [reply]
> Does your cc2538dk board have a enc28j60 device?

Yes, it is.The problem now is that the compilation fails.

**Reply by pjonsson** [reply]
Uncomment the `WITH_IP64 = 1` line that was commented out in the first step.

**Reply by tlzfighting** [reply]
### **Hi,sir!
     ###    I have uncommented the WITH_IP64 = 1 line that was commented out in the first step.
### **But when I type the compile command "make TARGET=cc2538dk."", the following error occurred.****

/usr/local/bin/../lib/gcc/arm-none-eabi/10.3.1/../../../../arm-none-eabi/bin/ld: build/cc2538dk/obj/enc28j60.o: in function `writereg':
enc28j60.c:(.text.writereg+0x6): undefined reference to `enc28j60_arch_spi_select'

### **How can I solve this problem for you, look forward to your valuable suggestions!**

**Reply by pjonsson** [reply]
Find the file that defines `enc28j60_arch_spi_select` and figure out why that isn't included in your build.

**Reply by tlzfighting** [reply]
The file that defines enc28j60_arch_spi_select has been included in " arch/dev/ethernet/enc28j60/enc28j60.h " .
But  I have typed MODULES += $(CONTIKI_NG_DRIVERS_ETHERNET_DIR)/enc28j60“” in the cc2538dk.Makefile.
 "enc28j60_arch_spi_select"can be found theoretically.
Or there is other way to use ip64-router on cc2538dk board?

**Reply by pjonsson** [reply]
That's the declaration. The definition is in one of these files:
```
arch/platform/zoul/orion/enc28j60-arch-gpio.c:enc28j60_arch_spi_select(void)
arch/platform/zoul/orion/enc28j60-arch-spi.c:enc28j60_arch_spi_select(void)
```

The comment on line 5 in ip64-conf.h explains the situation, the Orion board was the only hardware that had both 802.15.4 and ethernet that Contiki supported (when the example was added). It's possible your board could be supported, but nobody had tried it before you did, and the support clearly isn't in the tree right now.

**Reply by tlzfighting** [reply]
**Hi sir! There is other way to use Border router ( like ip64-router) on cc2538dk board?**

**Reply by joakimeriksson** [reply]
What do you want to do?

If it is only connect your 6lowpan network with the local Linux stack you can use the rpl-border-router on the board and tunslip6 to connect to the board (it will setup a tun-interface for IPv6 packets). In this case you can then reach your devices from your linux-machine and they can reach any application on the linux machine (but not reach Internet).

**Reply by tlzfighting** [reply]
Hi,sir!   I want to make my cc2538dk board  act as a border router.When this ### Border router is running, the data of other end nodes can be sent to the host (networks that use the same network segment as the Border router) through the ### Border router.

**Reply by joakimeriksson** [reply]
So you are not specifically interested in NAT64/DNS64 as provided by the ip64-router? Then you should probably look at rpl-border-router instead!

**Reply by tlzfighting** [reply]
> So you are not specifically interested in NAT64/DNS64 as provided by the ip64-router? Then you should probably look at rpl-border-router instead!

Hi ,sir! How can I use the function of  NAT64/DNS64  provided by the ip64-router  on cc2538dk board?

**Reply by tlzfighting** [reply]
> So you are not specifically interested in NAT64/DNS64 as provided by the ip64-router? Then you should probably look at rpl-border-router instead!

Follow the tutorial：RPL border router
First, program one or several nodes with hello-world (which has RPL enabled). Then, go to examples/rpl-border-router, compile and upload the example to the node you want to use as border router.
It is found that these nodes with hello-world (which has RPL enabled) are not connected to the border router's network. How to solve this problem? Looking forward to your valuable reply！
![图片](https://user-images.githubusercontent.com/95647449/209475880-26063167-3d74-4097-8e04-812716f6460f.png)
![图片](https://user-images.githubusercontent.com/95647449/209475882-7a87eae9-ce60-4bc5-b73a-c5fa3d5880c0.png)

**Reply by joakimeriksson** [reply]
Are you running all devices on the same platform - and did you check that they all are configured the same? (radio channel, etc).?

They should be discovered and end up in the router's tables.

**Reply by Mbaron12** [reply]
> Are you running all devices on the same platform - and did you check that they all are configured the same? (radio channel, etc).?
> 
> They should be discovered and end up in the router's tables.

Hi ,sir!  These devices  on the same cc2538dk platform can automatically join the border router when they work properly under the contiki 3.0 system, but they can not automatically join the border router's network under the contki-ng system.
These devices are configured as follows：
![image](https://user-images.githubusercontent.com/120808140/209553741-ffa9a316-8a6a-49a0-8a22-d856be8d1e36.png)

**Reply by Mbaron12** [reply]
> So you are not specifically interested in NAT64/DNS64 as provided by the ip64-router? Then you should probably look at rpl-border-router instead!

Hi,sir!
In order to use IP64DHCP , I have defined IP64DHCP in the arch/platform/cc2538dk/ip64-conf.h:

`#define IP64_CONF_DHCP                   1`
However, when I run the ip64-router-border on cc2538, IP64DHCP does not work and the serial port prints the following message：
```

[INFO: Main      ] Starting Contiki-NG-release/v4.8-533-g7a9349a57-dirty
[INFO: Main      ] - Routing: RPL Lite
[INFO: Main      ] - Net: sicslowpan
[INFO: Main      ] - MAC: CSMA
[INFO: Main      ] - 802.15.4 PANID: 0xabcd
[INFO: Main      ] - 802.15.4 Default channel: 26
[INFO: Main      ] Node ID: 18000
[INFO: Main      ] Link-layer address: 0012.4b00.18eb.4650
[INFO: Main      ] Tentative link-local IPv6 address: fe80::212:4b00:18eb:4650
[INFO: CC2538DK  ] TI SmartRF06 + cc2538EM
MAC addr 02:12:4b:eb:46:50
IP64: Starting DHCPv4
```
This only shows “IP64: Starting DHCPv4”.How can I solve this problem, I look forward to your valuable reply！
